django_lightweight_queue/app_settings.py

```diff
@@ -1,52 +1,177 @@
-from typing import Dict, Union, Mapping, TypeVar, Callable, Optional, Sequence
+from typing import Union, Mapping, TypeVar, Callable, Optional, Sequence
 
-from django.conf import settings
+from django.conf import settings as django_settings
 
 from . import constants
 from .types import Logger, QueueName
 
 T = TypeVar('T')
 
 
-def setting(suffix: str, default: T) -> T:
-    attr_name = '{}{}'.format(constants.SETTING_NAME_PREFIX, suffix)
-    return getattr(settings, attr_name, default)
+class Settings():
+    def _get(self, suffix: str, default: T) -> T:
+        attr_name = '{}{}'.format(constants.SETTING_NAME_PREFIX, suffix)
+        return getattr(django_settings, attr_name, default)
 
+    # adjustable values at runtime
+    _workers = None
+    _backend = None
+    _logger_factory = None
+    _backend_overrides = None
+    _middleware = None
+    _ignore_apps = None
+    _redis_host = None
+    _redis_port = None
+    _redis_password = None
+    _redis_prefix = None
+    _enable_prometheus = None
+    _prometheus_start_port = None
+    _atomic_jobs = None
 
-WORKERS = setting('WORKERS', {})  # type: Dict[QueueName, int]
-BACKEND = setting(
-    'BACKEND',
-    'django_lightweight_queue.backends.synchronous.SynchronousBackend',
-)  # type: str
+    @property
+    def WORKERS(self):
+        if not self._workers:
```

Comment:
Since these can be modified (as you noted, though I had forgotten!) and some of them can have legitimately `None` values, I'm concerned that this check (or even an `is not None` check) could result in incorrect overrides behaviour in some cases.

One option might be to introduce a concept of settings layers -- the defaults at the bottom, then the Django settings and finally an optional layer from the custom configuration specified on the command line. This might need changes in `load_extra_config` too, depending how it works, though I think that's ok if required.

Reply:
I've been thinking about this for the last few days and I think I have a simple solution that is the best of all worlds. Instead of checking to see if the cache value is falsey, we can swap to a dict in the Settings class and check to see if the cache value exists at all:

```python
_cache = {}

def WORKERS(self):
    if "WORKERS" not in self._cache:
        self._cache = self._get('WORKERS', self.get_empty_dict())
    return self._cache.get("WORKERS")
```

That way even if the cache returns a falsey value, the fact that it exists in the dict means that it has been called at least once so we can just return whatever we have stored and retain the intended capability.

What do you think?

Reply:
I think I follow what you're suggesting, however I'm not sure I understand how that translates to your example code. In particular I'm not sure I follow what role `self._cache` plays nor why it's needed (perhaps there are just errors in the code?).

I think something like that could work for sure.

Just as an alternative, fleshing out what I'd meant by layers
``` python
class Settings(Protocol):
    WORKERS: dict[...]
    ...

class Defaults:
    WORKERS = {}
    ...

class AppSettings:
    def __init__(self, layers: list[Settings]):
        self._layers = layers

    def add_layer(self, layer):  # to be called by `load_extra_config`
        self._layers.append(layer)

    def __getattr__(self, name):  # or maybe repeated per property, not sure
        # reverse so that later layers override earlier ones
        for layer in reversed(self._layers):
            if hasattr(layer, name):
                return getattr(layer, name)
        raise AttributeError(...)

app_settings: Settings = AppSettings(layers=[Defaults(), django_settings])
```

Not sure if this is any better than the dictionaries approach though (given that Python objects are approximately dictionaries anyway it's eventually similar in behaviour) and I don't feel strongly which approach we take.

As much as possible it would be great to keep the type checking useful, though I think it's ok to limit that just to the interface to the settings if that makes things simpler. (My gut feel is that we'll get slightly more out of them internally from the layers approach, but even then only as far as the `Defaults` class).

Similarly I'm not too worried about maintaining the precise semantics of the settings being writable. I don't think that was ever intended to be a formal part of the system, so as long as `load_extra_config` works [^1] it's probably fine.

[^1]: I ended up adding some basic tests for it in https://github.com/thread/django-lightweight-queue/pull/66, but don't worry about conflicts or anything - I doubt that'll get merged for a while

Reply:
Ah, okay interesting. My suggestion directly addresses the "what if we detect we've loaded the value but the value is None" question, but I think yours is a bit more straightforward. I'll poke at it a bit :+1: